lib/rules/no-trivially-nested-quantifier.ts

```diff
@@ -0,0 +1,282 @@
+        //     is a hole.
+        //
+        // We've shown that if there is no hole h between the first and second interval, then there can be no other
+        // holes. Therefore it is sufficient to only check for the first hole.
+
+        const a = child.min
+        const b = child.max
+        const c = parent.min
+        const d = parent.max
+        const condition =
+            b === Infinity && c === 0
+                ? a <= 1
+                : c === d || b * c + 1 >= a * (c + 1)
+
+        if (condition) {
+            return {
+                min: a * c,
+                max: b * d,
+                greedy,
+            }
+        }
+        return null
+    }
+    return null
+}
+
+/**
+ * Given a parent quantifier and a child quantifier, this will return a
+ * simplified child quant.
+ */
+function getSimplifiedChildQuant(
+    parent: Quantifier,
+    child: Quantifier,
+): Quant | null {
+    if (parent.max === 0 || child.max === 0) {
+        // this rule doesn't handle this
+        return null
+    } else if (parent.greedy !== child.greedy) {
+        // maybe some optimization is possible, but I'm not sure, so let's be safe
+        return null
+    }
+    let min = child.min
+    let max = child.max
+
+    if (min === 0 && parent.min === 0) {
+        min = 1
+    }
+    if (parent.max === Infinity && (min === 0 || min === 1) && max > 1) {
+        max = 1
+    }
+
+    return { min, max, greedy: child.greedy }
+}
+
+export default createRule("no-trivially-nested-quantifier", {
+    meta: {
+        docs: {
+            description:
+                "disallow nested quantifiers that can be rewritten as one quantifier",
+            // TODO Switch to recommended in the major version.
+            // recommended: true,
+            recommended: false,
+        },
+        fixable: "code",
+        schema: [],
+        messages: {
+            nested:
+                "These two quantifiers are trivially nested and can be replaced with '{{quant}}'.",
+            childOne: "This nested quantifier can be removed.",
+            childSimpler:
+                "This nested quantifier can be simplified to '{{quant}}'.",
+        },
+        type: "suggestion", // "problem",
+    },
+    create(context) {
+        const sourceCode = context.getSourceCode()
+
+        /**
+         * Create visitor
+         * @param node
+         */
+        function createVisitor(node: Expression): RegExpVisitor.Handlers {
+            const ignore: Set<RegExpNode> = new Set()
+
+            return {
+                onQuantifierEnter(qNode) {
+                    if (hasSomeAncestor(qNode, (a) => ignore.has(a))) {
+                        return
+                    }
+                    if (qNode.max === 0) {
+                        // this rule does not handle this case
+                        ignore.add(qNode)
+                        return
+                    }
```

Comment:
I don't think you need to check it ancestors if you use a Leave event.

```ts
            let ignore: Quantifier | null = null

            return {
                onQuantifierLeave(qNode) {
                    if (ignore === qNode) {
                        ignore = null
                    }
                },
                onQuantifierEnter(qNode) {
                    if (ignore) {
                        return
                    }
                    if (qNode.max === 0) {
                        // this rule does not handle this case
                        ignore = qNode
                        return
                    }
                    // ...
                }
```

Reply:
I don't think we need the `ignore` set at all, now that I think about it. I believe that I added this to avoid conflicts. (This was before I understood how fix conflicts really worked.)